apps/settings/js/modules/base/event_emitter.js

```diff
@@ -0,0 +1,127 @@
+define(function(require) {
+  'use strict';
+
+  var Module = require('modules/base/module');
+
+  /**
+   * EventEmitter implements basic event emitting logics. It also provides
+   * function that augments the ability to other objects.
+   *
+   * @example
+   *   var NewModule = Module.create(function() {
+   *     this.super(EventEmitter).call(this, ['event1', 'event2']);
+   *   }).extend(EventEmitter);
+   *
+   *   var module = NewModule();
+   *   var eventHandler = function(event) {
+   *     console.log(event.detail);
+   *   };
+   *   var obj = {
+   *     handleEvent: function(event) {
+   *       console.log(event.detail);
+   *     }
+   *   };
+   *
+   *   module.addEventListener('event1', eventHandler);
+   *   module.addEventListener('event2', handleEvent);
+   *
+   *   module._emitEvent('event1', 'value1'); // value1
+   *   module._emitEvent('event2', 'value2'); // value2
+   *
+   * @class EventEmitter
+   * @returns {EventEmitter}
+   */
+  var EventEmitter = Module.create(function EventEmitter(eventNames) {
+    if (eventNames && eventNames.length) {
+      this._eventListeners = eventNames.reduce((result, eventName) => {
+        result[eventName] = [];
+        return result;
+      }, {});
+    } else {
+      this.throw('no valid registered events');
+    }
+  });
+
+  /**
+   * Emit an event of a specific type and a given value.
+   *
+   * @access private
+   * @memberOf EventEmitter.prototype
+   * @param {String} eventName
+   * @param {Object} value
+   */
+  EventEmitter.prototype._emitEvent = function(eventName, value) {
+    var listeners = this._eventListeners[eventName];
+    if (!listeners) {
+      this.throw('invalid event name: ' + eventName);
+    }
+
+    this.debug('_emitEvent:' + eventName + ' ' + value);
+    var eventObj = {
+      type: eventName,
+      detail: value
+    };
+    listeners.forEach((listener) => {
+      listener.call(this, eventObj);
+    });
+  };
+
+  /**
+   * Add an event listener on an event. If listener is an object,
+   * object.handleEvent is called when the event emits.
+   *
+   * @access public
+   * @memberOf EventEmitter.prototype
+   * @param {String} eventName
+   * @param {(Function|Object)} listener
+   */
+  EventEmitter.prototype.addEventListener = function(eventName, listener) {
+    var listeners = this._eventListeners[eventName];
+    if (listener && listeners && listeners.indexOf(listener) < 0) {
```

Comment:
Per the offline discussion the event types should be predefined instead of run-time determined, so we still need the check.